url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/NestAgent.java

```diff
@@ -0,0 +1,363 @@
+	}
+
+	@Override
+	public void msgReceivePartDone() {
+		print("Received msgReceivePartDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgGivePartToPartsRobotDone() {
+		print("Received msgGivePartToPartsRobotDone from graphics");
+		animation.release();
+		stateChanged();
+	}
+
+	@Override
+	public void msgPurgingDone() {
+		print("Received msgPurgingDone from graphics");
+		state = NestState.DONE_PURGING;	
+		if(currentPartType == null) {
+			state = NestState.NULL;
+		}
+		animation.release();
+		stateChanged();
+	}
+	
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		// print("In scheduler");
+		if (state == NestState.PURGING || state == NestState.PRIORITY_PURGE) {
+			purgeSelf();
+			return true;
+		} else if (state == NestState.DONE_PURGING && laneState == LaneState.READY && !takingParts
+				&& currentPartType != null) {
+			if (partReady && currentParts.size() < full) {
+				requestPart();
+				return true;
+			}
+			synchronized (requestList) {
+				for (PartType requestedPart : requestList) {
+					if (count + countRequest < full) {
+						getParts(requestedPart);
+						return true;
+					}
+				}
+			}
+			synchronized (currentParts) {
+				for (MyPart currentPart : currentParts) {
+					if (currentPart.status == NestStatus.REMOVING) {
+						removePart(currentPart);
+						return true;
+					}
+				}
+			}
+			synchronized (currentParts) {
+				for (MyPart currentPart : currentParts) {
+					if (currentPart.status == NestStatus.IN_NEST) {
+						moveToPosition(currentPart);
+						return true;
+					}
+				}
+			}
+			if (currentParts.size() >= full && takingParts == false) {
+				nestFull();
+				return true;
+			}
+		}
+		return false;
+	}
+
+	// ACTIONS
+	public void purgeSelf() {
+		state = NestState.WAITING_FOR_LANE_PURGE;
+		if (nestGraphics != null) {
+			nestGraphics.purge();
+			try {
+				animation.acquire();
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+		}
+		takingParts = false;
+		requestList = Collections.synchronizedList(new ArrayList<PartType>());
+		currentParts = Collections.synchronizedList(new ArrayList<MyPart>());
+		countRequest = 0;
+		count = 0;
+		requestList.clear();
+		currentParts.clear();
+		requestList.add(currentPartType);
+		stateChanged();
+	}
+
+	public void requestPart() {
+		partReady = false;
+		lane.msgGiveMePart();
+		stateChanged();
+	}
+
+	public void getParts(PartType requestedType) {
+		print("Telling lane it need a part and incrementing count");
+		countRequest++;
+		lane.msgINeedPart(requestedType);
+
+		stateChanged();
+	}
+
+	public void moveToPosition(MyPart mp) {
+		print("Moving part to proper nest location");
+		// count++;
+		if (nestGraphics != null) {
+			// TODO
+			PartGraphics pg = new PartGraphics(mp.part.partGraphics);
+			nestGraphics.receivePart(mp.part.partGraphics);
+			try {
+				animation.acquire();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)